File: skills/skill-audit-ts/scripts/validate_agent_tool.ts
#!/usr/bin/env -S npx tsx
/**
 * Validate that a TypeScript script meets the agent tool interface contract.
 *
 * Checks: parseArgs (or another arg parser) present, --format/--quiet flags,
 * reachable exit codes 0/1/2/3, no interactive prompts, no stdout/stderr
 * mixing, non-builtin imports declared in a reachable package.json.
 *
 * Usage:
 *   validate_agent_tool.ts <script-path> [--json] [--exit-on-warn]
 *
 * Exit codes:
 *   0   All checks pass
 *   1   One or more FAIL-level issues found
 *   2   Bad invocation (path missing, unreadable)
 */

import { existsSync, readFileSync } from "node:fs";
import { builtinModules } from "node:module";
import { dirname, join, resolve } from "node:path";
import { parseArgs } from "node:util";
import ts from "typescript";

const EXIT_OK = 0;
const EXIT_FAIL = 1;
const EXIT_BAD_INVOCATION = 2;

export interface Finding {
  code: string;
  severity: "fail" | "warn" | "info";
  message: string;
  line?: number;
}

function ast(src: string): ts.SourceFile {
  return ts.createSourceFile("tool.ts", src, ts.ScriptTarget.Latest, true, ts.ScriptKind.TS);
}

function lineOf(sourceFile: ts.SourceFile, pos: number): number {
  return sourceFile.getLineAndCharacterOfPosition(pos).line + 1;
}

function forEachNode(node: ts.Node, visit: (n: ts.Node) => void): void {
  visit(node);
  ts.forEachChild(node, (child) => forEachNode(child, visit));
}

/**
 * Resolve the possible numeric exit-code values of an expression: a literal,
 * a named EXIT_* constant, or either branch of a ternary (idiomatic for
 * `return cond ? 1 : 0`). Returns all statically-derivable values.
 */
function resolveExitValues(
  arg: ts.Expression | undefined,
  constMap: Map<string, number>,
): number[] {
  if (!arg) return [];
  if (ts.isNumericLiteral(arg)) return [Number(arg.text)];
  if (ts.isIdentifier(arg)) {
    const val = constMap.get(arg.text);
    return val !== undefined ? [val] : [];
  }
  if (ts.isConditionalExpression(arg)) {
    return [
      ...resolveExitValues(arg.whenTrue, constMap),
      ...resolveExitValues(arg.whenFalse, constMap),
    ];
  }
  if (ts.isParenthesizedExpression(arg)) {
    return resolveExitValues(arg.expression, constMap);
  }
  return [];
}

/** Map every `const NAME = <numeric literal>` declaration in the file, for resolving EXIT_* constants. */
function buildNumericConstMap(sourceFile: ts.SourceFile): Map<string, number> {
  const constMap = new Map<string, number>();
  forEachNode(sourceFile, (node) => {
    if (
      ts.isVariableDeclaration(node) &&
      ts.isIdentifier(node.name) &&
      node.initializer &&
      ts.isNumericLiteral(node.initializer)
    ) {
      constMap.set(node.name.text, Number(node.initializer.text));
    }
  });
  return constMap;
}

function isProcessExitCall(node: ts.Node): node is ts.CallExpression {
  return (
    ts.isCallExpression(node) &&
    ts.isPropertyAccessExpression(node.expression) &&
    ts.isIdentifier(node.expression.expression) &&
    node.expression.expression.text === "process" &&
    node.expression.name.text === "exit"
  );
}

/** Every statically-derivable exit-code value reachable via `process.exit(...)` or `return ...`. */
function collectExitValues(sourceFile: ts.SourceFile, constMap: Map<string, number>): number[] {
  const values: number[] = [];
  forEachNode(sourceFile, (node) => {
    if (isProcessExitCall(node)) {
      values.push(...resolveExitValues(node.arguments[0], constMap));
    } else if (ts.isReturnStatement(node) && node.expression) {
      values.push(...resolveExitValues(node.expression, constMap));
    }
  });
  return values;
}

// ---------------------------------------------------------------------------
// Individual checks
// ---------------------------------------------------------------------------

function checkArgParser(src: string): Finding[] {
  if (
    !/\bparseArgs\s*\(/.test(src) &&
    !/\bcommander\b/i.test(src) &&
    !src.includes("new Command(")
  ) {
    return [
      {
        code: "contract.missing-argparse",
        severity: "fail",
        message: "No parseArgs()/commander call found — agents cannot call --help or pass flags",
      },
    ];
  }
  return [];
}

function checkFormatFlag(src: string): Finding[] {
  if (!src.includes("--format")) {
    return [
      {
        code: "contract.missing-format-flag",
        severity: "fail",
        message: "No --format flag — agents cannot request JSON output explicitly",
      },
    ];
  }
  return [];
}

function checkQuietFlag(src: string): Finding[] {
  if (!src.includes("--quiet")) {
    return [
      {
        code: "contract.missing-quiet-flag",
        severity: "warn",
        message: "No --quiet flag — agents cannot suppress informational stderr",
      },
    ];
  }
  return [];
}

function checkExitCodes(src: string): Finding[] {
  const sourceFile = ast(src);
  const constMap = buildNumericConstMap(sourceFile);
  const exitsFound = new Set(collectExitValues(sourceFile, constMap));

  // Contract (assets/templates/script.ts.template): 0 success, 1 issues found,
  // 2 bad invocation, 3 nothing matched. Every tool needs a success path and a
  // bad-invocation path. Codes 1 and 3 only mean something for tools that can
  // report findings or return an empty result — a converter or a counter has
  // neither, so flagging their absence as a warning is noise, not signal.
  const CHECKS: { code: number; severity: Finding["severity"]; message: string }[] = [
    { code: 0, severity: "warn", message: "Exit code 0 not found — success path missing" },
    {
      code: 2,
      severity: "warn",
      message: "Exit code 2 not found — bad-invocation path missing (unparseable args, bad path)",
    },
    {
      code: 1,
      severity: "info",
      message:
        "No exit code 1 — if this tool reports findings, return 1 when any are found so " +
        "agents can branch on the result without parsing output",
    },
    {
      code: 3,
      severity: "info",
      message:
        "No exit code 3 — if this tool can legitimately return zero results, return 3 so " +
        "agents can distinguish 'nothing matched' from 'error'",
    },
  ];
  return CHECKS.filter((c) => !exitsFound.has(c.code)).map(({ code, severity, message }) => ({
    code: `contract.missing-exit-${code}`,
    severity,
    message,
  }));
}

function checkNoInteractive(src: string): Finding[] {
  const sourceFile = ast(src);
  const findings: Finding[] = [];
  forEachNode(sourceFile, (node) => {
    if (!ts.isCallExpression(node)) return;
    const callee = node.expression;
    const isQuestionCall = ts.isPropertyAccessExpression(callee) && callee.name.text === "question";
    const isBarePromptCall = ts.isIdentifier(callee) && callee.text === "prompt";
    if (isQuestionCall || isBarePromptCall) {
      findings.push({
        code: "contract.interactive-prompt",
        severity: "fail",
        message: `Line ${lineOf(sourceFile, node.getStart(sourceFile))}: interactive prompt detected — agents cannot respond`,
        line: lineOf(sourceFile, node.getStart(sourceFile)),
      });
    }
  });
  return findings;
}

function checkStderrForErrors(src: string): Finding[] {
  const findings: Finding[] = [];
  const lines = src.split("\n");
  lines.forEach((line, i) => {
    const stripped = line.trim();
    if (
      stripped.startsWith("console.log(") &&
      stripped.includes("JSON.stringify") &&
      /error/i.test(stripped) &&
      !stripped.includes("process.stderr")
    ) {
      findings.push({
        code: "contract.error-on-stdout",
        severity: "warn",
        message: `Line ${i + 1}: possible error JSON on stdout — errors must go to stderr`,
        line: i + 1,
      });
    }
  });
  return findings;
}

const NODE_BUILTINS = new Set(builtinModules);

function isBuiltinModule(spec: string): boolean {
  const bare = spec.startsWith("node:") ? spec.slice(5) : spec;
  return spec.startsWith("node:") || NODE_BUILTINS.has(bare);
}

function extractImportedModules(src: string): string[] {
  const modules = new Set<string>();
  const importRe = /(?:import\s+(?:[\w*${},\s]+\s+from\s+)?|require\()\s*['"]([^'"]+)['"]/g;
  for (const m of src.matchAll(importRe)) {
    const spec = m[1]!;
    if (spec.startsWith(".") || spec.startsWith("/")) continue;
    modules.add(spec);
  }
  return [...modules];
}

function findNearestPackageJson(startDir: string): string | null {
  let dir = resolve(startDir);
  for (let i = 0; i < 20; i++) {
    const candidate = join(dir, "package.json");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return null;
}

function packageName(spec: string): string {
  // Scoped packages (@scope/pkg/subpath) keep the scope+name; unscoped
  // packages (pkg/subpath) keep only the leading segment.
  const parts = spec.split("/");
  if (spec.startsWith("@")) return parts.slice(0, 2).join("/");
  return parts[0]!;
}

function checkPackageDeps(src: string, scriptPath: string): Finding[] {
  const nonBuiltin = extractImportedModules(src)
    .map(packageName)
    .filter((m) => !isBuiltinModule(m));
  if (nonBuiltin.length === 0) return [];

  const pkgPath = findNearestPackageJson(dirname(resolve(scriptPath)));
  if (!pkgPath) {
    return [
      {
        code: "contract.missing-package-json",
        severity: "warn",
        message:
          `Non-builtin imports detected (${[...new Set(nonBuiltin)].sort().join(", ")}) but no ` +
          "reachable package.json — add one declaring these as dependencies so `npm ci` " +
          "resolves them.",
      },
    ];
  }

  let declared = new Set<string>();
  try {
    const pkg = JSON.parse(readFileSync(pkgPath, "utf-8")) as Record<string, unknown>;
    for (const field of [
      "dependencies",
      "devDependencies",
      "optionalDependencies",
      "peerDependencies",
    ]) {
      const deps = pkg[field];
      if (deps && typeof deps === "object") {
        declared = new Set([...declared, ...Object.keys(deps)]);
      }
    }
  } catch {
    // Malformed package.json — treat as if nothing is declared.
  }

  const missing = [...new Set(nonBuiltin)].filter((m) => !declared.has(m)).sort();
  if (missing.length === 0) return [];
  return [
    {
      code: "contract.missing-package-dep",
      severity: "warn",
      message:
        `package.json at ${pkgPath} does not declare: ${missing.join(", ")} — add them as ` +
        "dependencies so `npm ci` resolves what this script imports.",
    },
  ];
}

function checkShebang(src: string): Finding[] {
  if (!src.startsWith("#!/usr/bin/env -S npx tsx") && !src.startsWith("#!/usr/bin/env node")) {
    return [
      {
        code: "contract.missing-shebang",
        severity: "info",
        message: "No '#!/usr/bin/env -S npx tsx' shebang — script won't be directly executable",
      },
    ];
  }
  return [];
}

function checkDocComment(src: string): Finding[] {
  const withoutShebang = src.replace(/^#!.*\n/, "");
  if (!/^\s*\/\*\*/.test(withoutShebang)) {
    return [
      {
        code: "contract.missing-docstring",
        severity: "info",
        message: "No leading /** doc comment — add one describing inputs, outputs, exit codes",
      },
    ];
  }
  return [];
}

function checkEpilog(src: string): Finding[] {
  if (/examples?:/i.test(src)) return [];
  return [
    {
      code: "contract.missing-epilog",
      severity: "warn",
      message:
        "No 'Examples:' text found in help output — `--help` is the agent's primary " +
        "reference; show 2-3 realistic invocations",
    },
  ];
}

function checkStructuredStderrErrors(src: string): Finding[] {
  const sourceFile = ast(src);
  const constMap = buildNumericConstMap(sourceFile);
  const hasNonZeroExit = collectExitValues(sourceFile, constMap).some((v) => v !== 0);

  if (!hasNonZeroExit) return [];

  if (src.includes("emitError(") || src.includes("emit_error(")) return [];
  for (const line of src.split("\n")) {
    const stripped = line.trim();
    if (
      stripped.includes("JSON.stringify") &&
      (stripped.includes("process.stderr") || stripped.includes("console.error")) &&
      /error/i.test(stripped)
    ) {
      return [];
    }
  }

  return [
    {
      code: "contract.missing-structured-stderr",
      severity: "warn",
      message:
        "Non-zero exit paths exist but no structured error JSON found on stderr — errors " +
        'should be `process.stderr.write(JSON.stringify({error, code, hint}) + "\\n")` so the ' +
        "agent gets a consistent shape",
    },
  ];
}

// ---------------------------------------------------------------------------
// Runner
// ---------------------------------------------------------------------------

const ALL_CHECKS: Array<(src: string) => Finding[]> = [
  checkArgParser,
  checkFormatFlag,
  checkQuietFlag,
  checkExitCodes,
  checkNoInteractive,
  checkStderrForErrors,
  checkShebang,
  checkDocComment,
  checkEpilog,
  checkStructuredStderrErrors,
];

export function validate(scriptPath: string): Finding[] {
  const src = readFileSync(scriptPath, "utf-8");
  const findings: Finding[] = [];
  for (const check of ALL_CHECKS) {
    findings.push(...check(src));
  }
  findings.push(...checkPackageDeps(src, scriptPath));
  return findings;
}

function emitText(path: string, findings: Finding[]): void {
  for (const f of findings) {
    const loc = f.line ? ` (line ${f.line})` : "";
    console.log(`${f.severity.toUpperCase()}: [${f.code}]${loc} ${f.message}`);
  }
  if (findings.length === 0) {
    console.log(`OK: ${path} passes all checks`);
  }
}

function emitJson(path: string, findings: Finding[]): void {
  console.log(
    JSON.stringify(
      {
        target: path,
        findings,
        summary: {
          total: findings.length,
          fail: findings.filter((f) => f.severity === "fail").length,
          warn: findings.filter((f) => f.severity === "warn").length,
        },
      },
      null,
      2,
    ),
  );
}

const HELP = `Validate that a TypeScript script meets the agent tool interface contract.

Usage:
  validate_agent_tool.ts <script-path> [--format json|text] [--exit-on-warn]

Examples:
  validate_agent_tool.ts tool.ts                  # text report to stdout
  validate_agent_tool.ts tool.ts --format json    # machine-readable JSON
  validate_agent_tool.ts tool.ts --exit-on-warn   # CI mode: warnings fail

Options:
  --format <json|text>  Output format (default: text)
  --json                 Alias for --format json
  --quiet                Suppress informational stderr (no effect on findings output)
  --exit-on-warn         Exit 1 on warnings as well as failures
  --help                 Show this help message
`;

export function main(argv: string[]): number {
  const { values, positionals } = parseArgs({
    args: argv,
    allowPositionals: true,
    options: {
      format: { type: "string", default: "text" },
      json: { type: "boolean", default: false },
      quiet: { type: "boolean", default: false },
      "exit-on-warn": { type: "boolean", default: false },
      help: { type: "boolean", default: false },
    },
  });

  if (values.help) {
    console.log(HELP);
    return EXIT_OK;
  }

  const scriptArg = positionals[0];
  if (!scriptArg) {
    console.error(HELP);
    return EXIT_BAD_INVOCATION;
  }

  const path = resolve(scriptArg);
  if (!existsSync(path)) {
    console.error(`error: path does not exist: ${path}`);
    return EXIT_BAD_INVOCATION;
  }

  const useJson = values.json || values.format === "json";
  const src = readFileSync(path, "utf-8");
  const head = src.split("\n").slice(0, 10).join("\n");
  if (head.includes("agent-tool: false")) {
    if (useJson) {
      console.log(JSON.stringify({ target: path, skipped: true }, null, 2));
    } else if (!values.quiet) {
      console.log(`SKIP: ${path} (marked '// agent-tool: false')`);
    }
    return EXIT_OK;
  }

  const findings = validate(path);
  if (useJson) {
    emitJson(path, findings);
  } else {
    emitText(path, findings);
  }

  const hasFail = findings.some((f) => f.severity === "fail");
  const hasWarn = findings.some((f) => f.severity === "warn");
  if (hasFail) return EXIT_FAIL;
  if (values["exit-on-warn"] && hasWarn) return EXIT_FAIL;
  return EXIT_OK;
}

const isMain = process.argv[1] && import.meta.url === `file://${process.argv[1]}`;
if (isMain) {
  process.exit(main(process.argv.slice(2)));
}

File: skills/agent-tool-builder/scripts/validate_agent_tool.py
#!/usr/bin/env python3
"""Validate that a Python script meets the agent tool interface contract.

Checks: argparse present, --format/--quiet flags, reachable exit codes 0/1/2,
no exits in the shell-reserved range, argparse's own exit 2 remapped, no input()
calls, no stdout/stderr mixing, PEP 723 block if non-stdlib imports detected.

Usage:
    validate_agent_tool.py <script-path> [--json] [--exit-on-warn]

Exit codes:
    0   All checks pass
    1   One or more FAIL-level issues found
    2   Bad invocation (path missing, unreadable)
"""

from __future__ import annotations

import argparse
import ast
from dataclasses import asdict, dataclass
import json
from pathlib import Path
import re
import sys
import textwrap

EXIT_OK = 0
EXIT_FAIL = 1
EXIT_BAD_INVOCATION = 2


@dataclass
class Finding:
    code: str
    severity: str  # "fail" | "warn" | "info"
    message: str
    line: int | None = None


def _src(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def _ast(src: str) -> ast.Module | None:
    try:
        return ast.parse(src)
    except SyntaxError:
        return None


# ---------------------------------------------------------------------------
# Individual checks
# ---------------------------------------------------------------------------


def check_argparse(src: str) -> list[Finding]:
    if "ArgumentParser" not in src:
        return [
            Finding(
                "contract.missing-argparse",
                "fail",
                "No argparse.ArgumentParser found — agents cannot call --help or pass flags",
            )
        ]
    return []


def check_format_flag(src: str) -> list[Finding]:
    if "--format" not in src:
        return [
            Finding(
                "contract.missing-format-flag",
                "fail",
                "No --format flag — agents cannot request JSON output explicitly",
            )
        ]
    return []


def check_quiet_flag(src: str) -> list[Finding]:
    if "--quiet" not in src:
        return [
            Finding(
                "contract.missing-quiet-flag",
                "warn",
                "No --quiet flag — agents cannot suppress informational stderr",
            )
        ]
    return []


def _exit_values(src: str) -> set[int] | None:
    """Statically-derivable exit codes: sys.exit(N)/exit(N) and `return N` in main().

    Returns None when the source does not parse.
    """
    tree = _ast(src)
    if tree is None:
        return None

    # Build a map of name → literal int value for constants like EXIT_OK = 0
    const_map: dict[str, int] = {}
    for node in ast.walk(tree):
        if (
            isinstance(node, ast.Assign)
            and len(node.targets) == 1
            and isinstance(node.targets[0], ast.Name)
            and isinstance(node.value, ast.Constant)
            and isinstance(node.value.value, int)
        ):
            const_map[node.targets[0].id] = node.value.value

    exits_found: set[int] = set()
    for node in ast.walk(tree):
        # sys.exit(N) and exit(N) with literal or named constant
        if isinstance(node, ast.Call):
            func = node.func
            is_sys_exit = (
                isinstance(func, ast.Attribute)
                and func.attr == "exit"
                and isinstance(func.value, ast.Name)
                and func.value.id == "sys"
            )
            is_bare_exit = isinstance(func, ast.Name) and func.id == "exit"
            if (is_sys_exit or is_bare_exit) and node.args:
                arg = node.args[0]
                if isinstance(arg, ast.Constant) and isinstance(arg.value, int):
                    exits_found.add(arg.value)
                elif isinstance(arg, ast.Name) and arg.id in const_map:
                    exits_found.add(const_map[arg.id])
        # return N or return EXIT_* inside main() — idiomatic when sys.exit(main())
        elif isinstance(node, ast.Return) and node.value is not None:
            val = node.value
            if isinstance(val, ast.Constant) and isinstance(val.value, int):
                exits_found.add(val.value)
            elif isinstance(val, ast.Name) and val.id in const_map:
                exits_found.add(const_map[val.id])

    return exits_found


def check_exit_codes(src: str) -> list[Finding]:
    exits_found = _exit_values(src)
    if exits_found is None:
        return []

    labels = {
        0: "success path missing",
        1: "user-error path missing",
        2: "system-error path missing",
    }
    findings = [
        Finding(f"contract.missing-exit-{n}", "warn", f"Exit code {n} not found — {labels[n]}")
        for n in (0, 1, 2)
        if n not in exits_found
    ]
    if 3 not in exits_found:
        findings.append(
            Finding(
                "contract.missing-exit-3",
                "warn",
                "Exit code 3 (not-found) not found — agents cannot distinguish "
                "'nothing matched' from 'error'",
            )
        )
    return findings


# Codes the shell and kernel produce on their own. 130 (SIGINT) is the one a script
# may legitimately return, since it only restates what the shell would have reported.
_RESERVED_EXITS = {
    126: "command invoked cannot execute",
    127: "command not found",
    137: "killed by SIGKILL or the OOM killer",
    139: "segmentation fault",
    143: "killed by SIGTERM",
    255: "exit status out of range",
}


def check_reserved_exit_codes(src: str) -> list[Finding]:
    exits_found = _exit_values(src)
    if exits_found is None:
        return []
    return [
        Finding(
            f"contract.reserved-exit-{n}",
            "warn",
            f"Exits {n}, reserved by the shell ({_RESERVED_EXITS.get(n, 'signal 128+n')}) — "
            "the agent cannot tell it from the tool failing to launch or being killed, "
            "and no structured stderr line reaches it. Use 1/2/3.",
        )
        for n in sorted(exits_found)
        if n >= 126 and n != 130
    ]


def check_argparse_exit_remap(src: str) -> list[Finding]:
    """Flag scripts where argparse's own exit 2 collides with their meaning for 2.

    argparse exits 2 on an unknown flag or a bad ``type=``. That only misleads an
    agent when the script gives 2 a different meaning — a system/infrastructure
    error, which the agent may retry. A script whose 2 already means "bad
    invocation" agrees with argparse and needs no remap.
    """
    tree = _ast(src)
    if tree is None or "ArgumentParser" not in src:
        return []

    collides = any(
        isinstance(node, ast.Assign)
        and len(node.targets) == 1
        and isinstance(node.targets[0], ast.Name)
        and isinstance(node.value, ast.Constant)
        and node.value.value == 2
        and re.search(r"SYSTEM|INFRA|TRANSIENT|INTERNAL", node.targets[0].id)
        for node in ast.walk(tree)
    )
    if not collides:
        return []

    for node in ast.walk(tree):
        if not isinstance(node, ast.ClassDef):
            continue
        bases = {
            b.attr if isinstance(b, ast.Attribute) else getattr(b, "id", "") for b in node.bases
        }
        if "ArgumentParser" in bases and any(
            isinstance(b, ast.FunctionDef) and b.name == "error" for b in node.body
        ):
            return []

    return [
        Finding(
            "contract.argparse-exit-2",
            "info",
            "Exit 2 means system error here, but argparse also exits 2 on an unknown flag or a "
            "bad --type value — so an unfixable invocation looks like a retryable outage, and "
            "the agent gets usage text instead of the structured stderr line. Subclass "
            "ArgumentParser and override error() to emit the error line and exit 1.",
        )
    ]


def check_no_interactive(src: str) -> list[Finding]:
    """Check for interactive prompts using AST — avoids false positives in docstrings/comments."""
    tree = _ast(src)
    if tree is None:
        return []

    def _call_name(node: ast.Call) -> str | None:
        if isinstance(node.func, ast.Name):
            return node.func.id
        if isinstance(node.func, ast.Attribute):
            return node.func.attr
        return None

    return [
        Finding(
            "contract.interactive-prompt",
            "fail",
            f"Line {node.lineno}: interactive prompt detected — agents cannot respond",
            line=node.lineno,
        )
        for node in ast.walk(tree)
        if isinstance(node, ast.Call) and _call_name(node) in {"input", "getpass"}
    ]


def check_stderr_for_errors(src: str) -> list[Finding]:
    # Heuristic: flag print(json.dumps(...error...)) on stdout — errors must go to stderr
    return [
        Finding(
            "contract.error-on-stdout",
            "warn",
            f"Line {i}: possible error JSON on stdout — errors must go to stderr",
            line=i,
        )
        for i, line in enumerate(src.splitlines(), 1)
        if (stripped := line.strip())
        and "json.dumps" in stripped
        and "error" in stripped.lower()
        and "file=sys.stderr" not in stripped
        and stripped.startswith("print(")
    ]


# sys.stdlib_module_names is exhaustive and ships with the interpreter (3.10+).
# We add __future__ explicitly because it's a pseudo-module not always listed.
_STDLIB_MODULES = set(sys.stdlib_module_names) | {"__future__"}

_PEP723_BLOCK_RE = re.compile(r"^# /// script\s*\n(.*?)^# ///\s*$", re.MULTILINE | re.DOTALL)
_IMPORT_RE = re.compile(r"^import (\w+)|^from (\w+)", re.MULTILINE)


def check_pep723(src: str) -> list[Finding]:
    non_stdlib_modules: set[str] = set()
    for match in _IMPORT_RE.findall(src):
        mod = match[0] or match[1]
        if mod and mod not in _STDLIB_MODULES:
            non_stdlib_modules.add(mod)
    stdlib_only = not non_stdlib_modules

    block_match = _PEP723_BLOCK_RE.search(src)
    findings: list[Finding] = []

    if not block_match:
        if not stdlib_only:
            findings.append(
                Finding(
                    "contract.missing-pep723",
                    "warn",
                    "Non-stdlib imports detected ("
                    + ", ".join(sorted(non_stdlib_modules))
                    + ") but no PEP 723 inline metadata block — add a '# /// script' "
                    "block so `uv run` resolves dependencies, or confirm this script "
                    "is part of a package whose pyproject.toml owns deps",
                )
            )
        return findings

    block = block_match.group(1)
    if not re.search(r"^#\s*requires-python\s*=", block, re.MULTILINE):
        findings.append(
            Finding(
                "contract.pep723-missing-requires-python",
                "warn",
                "PEP 723 block present but no 'requires-python' key — declare the "
                "minimum Python version your code actually uses (e.g. '>=3.11')",
            )
        )
    if not stdlib_only and not re.search(r"^#\s*dependencies\s*=", block, re.MULTILINE):
        findings.append(
            Finding(
                "contract.pep723-missing-dependencies",
                "warn",
                "PEP 723 block present but no 'dependencies' key, despite non-stdlib "
                "imports (" + ", ".join(sorted(non_stdlib_modules)) + ") — add them",
            )
        )
    return findings


def check_shebang(src: str) -> list[Finding]:
    if not src.startswith("#!/usr/bin/env python3"):
        return [
            Finding(
                "contract.missing-shebang",
                "info",
                "No '#!/usr/bin/env python3' shebang — script won't be directly executable",
            )
        ]
    return []


def check_docstring(src: str) -> list[Finding]:
    tree = _ast(src)
    if tree is None:
        return []
    if not (
        tree.body
        and isinstance(tree.body[0], ast.Expr)
        and isinstance(tree.body[0].value, ast.Constant)
    ):
        return [
            Finding(
                "contract.missing-docstring",
                "info",
                "No module docstring — add one describing inputs, outputs, exit codes",
            )
        ]
    return []


def _parser_ctors(tree: ast.Module) -> set[str]:
    """`ArgumentParser` plus any local subclass of it — e.g. a ContractParser."""
    names = {"ArgumentParser"}
    for node in ast.walk(tree):
        if isinstance(node, ast.ClassDef) and any(
            (b.attr if isinstance(b, ast.Attribute) else getattr(b, "id", "")) in names
            for b in node.bases
        ):
            names.add(node.name)
    return names


def check_epilog(src: str) -> list[Finding]:
    """`--help` is the agent's reference; a flag list without examples leaves it guessing."""
    tree = _ast(src)
    if tree is None:
        return []
    ctors = _parser_ctors(tree)
    for node in ast.walk(tree):
        if (
            isinstance(node, ast.Call)
            and (
                (isinstance(node.func, ast.Attribute) and node.func.attr in ctors)
                or (isinstance(node.func, ast.Name) and node.func.id in ctors)
            )
            and any(kw.arg == "epilog" for kw in node.keywords)
        ):
            return []
    return [
        Finding(
            "contract.missing-epilog",
            "warn",
            "ArgumentParser has no epilog= with usage examples — "
            "`--help` is the agent's primary reference; show 2-3 realistic invocations",
        )
    ]


def check_structured_stderr_errors(src: str) -> list[Finding]:
    """Tools that exit with user/system errors should emit structured JSON to stderr."""
    tree = _ast(src)
    if tree is None:
        return []

    has_nonzero_exit = False
    for node in ast.walk(tree):
        if isinstance(node, ast.Call):
            func = node.func
            is_exit = (isinstance(func, ast.Attribute) and func.attr == "exit") or (
                isinstance(func, ast.Name) and func.id == "exit"
            )
            if is_exit and node.args:
                arg = node.args[0]
                if isinstance(arg, ast.Constant) and isinstance(arg.value, int) and arg.value != 0:
                    has_nonzero_exit = True
                    break
                if isinstance(arg, ast.Name) and arg.id.startswith("EXIT_") and arg.id != "EXIT_OK":
                    has_nonzero_exit = True
                    break
        elif isinstance(node, ast.Return) and node.value is not None:
            val = node.value
            if isinstance(val, ast.Constant) and isinstance(val.value, int) and val.value != 0:
                has_nonzero_exit = True
                break
            if isinstance(val, ast.Name) and val.id.startswith("EXIT_") and val.id != "EXIT_OK":
                has_nonzero_exit = True
                break

    if not has_nonzero_exit:
        return []

    for line in src.splitlines():
        stripped = line.strip()
        if (
            "json.dumps" in stripped
            and "file=sys.stderr" in stripped
            and ("error" in stripped.lower() or "_emit_error" in stripped.lower())
        ):
            return []
    if "_emit_error" in src or "emit_error" in src:
        return []

    return [
        Finding(
            "contract.missing-structured-stderr",
            "warn",
            "Non-zero exit paths exist but no structured error JSON found on stderr — "
            'errors should be `print(json.dumps({"error": ..., "code": ..., "hint": ...}), '
            "file=sys.stderr)` so the agent gets a consistent shape",
        )
    ]


# ---------------------------------------------------------------------------
# Runner
# ---------------------------------------------------------------------------

ALL_CHECKS = [
    check_argparse,
    check_format_flag,
    check_quiet_flag,
    check_exit_codes,
    check_reserved_exit_codes,
    check_argparse_exit_remap,
    check_no_interactive,
    check_stderr_for_errors,
    check_pep723,
    check_shebang,
    check_docstring,
    check_epilog,
    check_structured_stderr_errors,
]


def validate(path: Path) -> list[Finding]:
    src = _src(path)
    findings: list[Finding] = []
    for check in ALL_CHECKS:
        findings.extend(check(src))
    return findings


def _emit_text(path: Path, findings: list[Finding]) -> None:
    for f in findings:
        loc = f" (line {f.line})" if f.line else ""
        print(f"{f.severity.upper()}: [{f.code}]{loc} {f.message}")
    if not findings:
        print(f"OK: {path} passes all checks")


def _emit_json(path: Path, findings: list[Finding]) -> None:
    print(
        json.dumps(
            {
                "target": str(path),
                "findings": [asdict(f) for f in findings],
                "summary": {
                    "total": len(findings),
                    "fail": sum(1 for f in findings if f.severity == "fail"),
                    "warn": sum(1 for f in findings if f.severity == "warn"),
                },
            },
            indent=2,
        )
    )


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description=__doc__.split("\n", 1)[0] if __doc__ else "",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=textwrap.dedent("""\
            examples:
              validate_agent_tool.py tool.py                  # text report to stdout
              validate_agent_tool.py tool.py --format json    # machine-readable JSON
              validate_agent_tool.py tool.py --exit-on-warn   # CI mode: warnings fail
        """),
    )
    parser.add_argument("script", help="Path to the Python script to validate")
    parser.add_argument(
        "--format",
        choices=["json", "text"],
        default="text",
        help="Output format (default: text). The validator's audience is humans by default; "
        "use --format json from other tools.",
    )
    parser.add_argument(
        "--json",
        dest="as_json",
        action="store_true",
        help="Alias for --format json (kept for back-compat).",
    )
    parser.add_argument(
        "--quiet",
        action="store_true",
        help="Suppress informational stderr (no effect on findings output).",
    )
    parser.add_argument(
        "--exit-on-warn", action="store_true", help="Exit 1 on warnings as well as failures"
    )
    args = parser.parse_args(argv)

    path = Path(args.script).expanduser().resolve()
    if not path.exists():
        print(f"error: path does not exist: {path}", file=sys.stderr)
        return EXIT_BAD_INVOCATION

    # Skip library modules that opt out of agent-tool validation.
    src = _src(path)
    head = "\n".join(src.splitlines()[:10])
    if "# agent-tool: false" in head:
        use_json = args.as_json or args.format == "json"
        if use_json:
            print(json.dumps({"target": str(path), "skipped": True}, indent=2))
        elif not args.quiet:
            print(f"SKIP: {path} (marked # agent-tool: false)")
        return EXIT_OK

    findings = validate(path)

    use_json = args.as_json or args.format == "json"
    if use_json:
        _emit_json(path, findings)
    else:
        _emit_text(path, findings)

    has_fail = any(f.severity == "fail" for f in findings)
    has_warn = any(f.severity == "warn" for f in findings)
    if has_fail:
        return EXIT_FAIL
    if args.exit_on_warn and has_warn:
        return EXIT_FAIL
    return EXIT_OK


if __name__ == "__main__":
    sys.exit(main())
